### DISCUSSION TITLE
Debugging hooks/full-custom scripts

### DISCUSSION THREAD

**Discussion by Jesseeee** [rauc/rauc#1337] [Q&A] [answered]
We have some use cases where a hook or full-custom update is necessary, we'd like to add some logging in there so we can verify our steps happen as expected, what we're struggling with is how we could expose this during the install process like a regular update would notify of it's progress.
A standard echo doesn't show anything, we can write logging to a file but that's a bit cumbersome and would much rather have a way to implement this directly in our scripting for this if anyone knows how we could achieve this.
Thanks!

**Comment by ejoerns** [comment, marked as answer]
@Jesseeee The hook output will be visible in the service log. So if using systemd, it should end up in the journal. But I assume this isn't sufficient for your use case?

**Reply by OnkelUlla** [reply]
@Jesseeee As a quick hack you could redirect your output to ``/dev/console`` if the system has consoles.

**Reply by Jesseeee** [reply]
Well now I feel stupid for asking, luckily physical facepalms don't translate on the internet.
And yes for our current use case this will suffice as it's mostly following along with the install process to see it's going well.
But looking ahead, what could be useful for us is still showing this to the output when executing the rauc install command, as then we could provide some custom messages we could then filter out and show to the user.
A use case I'm thinking of is that we have a lot of custom settings for our variety of customers and different part numbers even between the same device and so on, so we might want to add checks in the script to ensure these are being applied to the correct device, easily getting output from this that can indicate e.g. an incorrect device, version or other parameter in the system would be very helpful.
I'm open to any suggestions of what an easy logging/feedback system in the scripts that doesn't go to the journal could look like :)

Thanks for the help!

**Reply by ejoerns** [reply]
@Jesseeee Thank you for sharing your use case!

From what I read I assume that you intend to use an [install-check hook](https://rauc.readthedocs.io/en/latest/using.html#install-hooks)?

It does not provide some sort of custom message interface, but it allows at least, in case of a failure, to return the last error message as installation error output.

Since the output you see in the install client or via D-Bus is tightly coupled to the progress reporting in RAUC, adding custom messages is not really supported yet in RAUC.

But generally thinking about an interface for providing custom installation output reminds me of what we have implemented for *handlers* in https://github.com/rauc/rauc/blob/master/src/install.c#L501 a while ago.
It was actually meant to be used for the special *full custom* handler, but it seems to work for normal handlers (like `pre-install`), too. Hooks however don't have a similar parsing and the messages won't end up in the 'end-user' output anyway. Thus I am unsure if this approach fits here.
